File: Lecture 1 - C/Problem Set 1/Mario-More/mario.c
/*
    PROBLEM: MARIO (MORE COMFORTABLE) - STANDARD C VERSION

    Write a program that:
    - Prompts the user for a height (1–8).
    - Prints a double half-pyramid.
    - No CS50 library used.
*/

#include <stdio.h>

// Function prototypes
int get_height(void);
void print_space(int space);
void print_rows(int row);

int main(void)
{
    int height = get_height();

    for (int i = 0; i < height; i++)
    {
        print_space(height - (i + 1));
        print_rows(i + 1);
    }

    return 0;
}

// Get height using standard input
int get_height(void)
{
    int height;

    do
    {
        printf("Height: ");
        scanf("%d", &height);
    }
    while (height < 1 || height > 8);

    return height;
}

// Print spaces
void print_space(int space)
{
    for (int i = 0; i < space; i++)
    {
        printf(" ");
    }
}

// Print pyramid rows
void print_rows(int row)
{
    for (int i = 0; i < row; i++)
    {
        printf("#");
    }

    printf("  ");

    for (int i = 0; i < row; i++)
    {
        printf("#");
    }

    printf("\n");
}

/*
---------------- ALGORITHM ----------------

1. Prompt user for integer between 1 and 8.
2. Repeat until valid input.
3. For each row:
    a. Print decreasing spaces.
    b. Print increasing hashes.
    c. Print two-space gap.
    d. Print increasing hashes again.
4. Move to next line.
5. End.

Time Complexity: O(n²)
Space Complexity: O(1)
*/

File: Lecture 6 - Python/Problem Set 6/Mario - More/mario.py
"""
PROBLEM: MARIO (MORE COMFORTABLE)

Write a program that:
1. Prompts the user for a height (integer between 1 and 8).
2. Prints a double half-pyramid of that height.
3. The two pyramids are separated by two spaces.
"""

# Keep asking until valid input is given
while True:
    try:
        # Take input from user and convert to integer
        height = int(input("Height: "))

        # Check if height is between 1 and 8
        if 1 <= height <= 8:
            break
        else:
            print("Give me integer between 1 to 8")

    # If user enters non-integer value
    except ValueError:
        print("Give me a integer between 1 to 8")


# Print the pyramid
for i in range(height):

    # Print leading spaces
    for j in range(height - (i + 1)):
        print(" ", end="")

    # Print left hashes
    for j in range(i + 1):
        print("#", end="")

    # Print gap between pyramids
    print("  ", end="")

    # Print right hashes
    for j in range(i + 1):
        print("#", end="")

    # Move to next line after each row
    print()


"""
---------------- ALGORITHM ----------------

1. Repeat forever:
    a. Try to take integer input from user.
    b. If input is between 1 and 8:
         break loop.
    c. Otherwise:
         print error message.
    d. If input is not an integer:
         catch ValueError and print error message.

2. For each row i from 0 to height - 1:
    a. Print (height - i - 1) spaces.
    b. Print (i + 1) hashes.
    c. Print two spaces.
    d. Print (i + 1) hashes.
    e. Move to next line.

Time Complexity: O(n²)
Space Complexity: O(1)
"""